design/baremetal-operator/user-defined-parameters.md

```diff
@@ -0,0 +1,456 @@
+<!--
+ This work is licensed under a Creative Commons Attribution 3.0
+ Unported License.
+
+ http://creativecommons.org/licenses/by/3.0/legalcode
+-->
+
+# Add Support for User-defined Parameters
+
+## Status
+
+provisional
+
+## Summary
+
+Add an optional field `extraParametersName` to accept a `Secret` object
+where to allow user to set vendor specific configuration.
+
+## Motivation
+
+When deploying BM nodes, some vendor's BMC will require additional parameters
+to adapt to certain situations. For example:
+
+1. When FIPS mode is enabled on the host running the ironic service, the
+   version of SNMP protocol of iRMC driver must be v3. Therefore, additional
+   parameters are needed to specify the version and to provide the corresponding
+   credential information.
+2. With some BMC firmware upgrades, additional configuration may be required.
+   E.g.: For iRMC firmware version >= `iRMC S6 2.00`, we need to additionally
+   provide driver info such as:
+    - **redifsh_address**: iRMC IP address or hostname.
+    - **redfish_username**: iRMC user name with administrative privileges.
+    - **redfish_password**: password of redfish_username.
+    - **redfish_verify_ca**: accepts values those accepted in irmc_verify_ca.
+3. For the production environment, it may be necessary to use a certificate
+   issued by a third-party CA for verification, so additional parameters are
+   required to specify the root certificate, e.g.: `irmc_verify_ca=<cert_path.crt>`.
```

Comment:
All we want to do is provide a parameter that can be passed to the ironic. As for the subsequent processing of this parameter or whether this parameter is validate, these are not in our proposal. We think these should be handled in `ironic`. I will also add this content to Non-goals.

In this example, we only consider passing the verify_ca path to ironic. As for the mounting and generation of the certificate, it should be handled by the user.

Reply:
> As for the subsequent processing of this parameter or whether this parameter is validate, these are not in our proposal.

This is the problem. BareMetalHost, for better or worse, is designed to be an independent API, not just a pass-thru to Ironic. We need to think in terms of this API. (As an aside, I also regret the way driver_info is designed in Ironic, but that ship has sailed long ago).

> As for the mounting and generation of the certificate, it should be handled by the user.

User in case of upstream Metal3 - sure. If you care about OpenShift, it's not possible for a user to modify the Metal3 pod in a supported way. This is not our limitation, this is how OpenShift works.

Reply:
Yeah, I was about to suggest the same. We already have this pattern for API TLS: define a fixed location where the certificates can be mounted. Then we also need a configuration in ironic.conf (if not already) that is set conditionally on this location existence. Then we update CBO (in case of OpenShift) to mount the secret with the certificates there. No BMO changes required.

Reply:
The thing I was in other places in OpenShift is putting your CA certificate as a string to install-config. That could be taken by the installer and put 1) on the bootstrap VM for control plane installation, 2) in the CBO configuration for the final cluster.

Getting certificate from a server causes a chicken-and-egg problem: how do we verify the certificate of that server? Also not all users will want to establish an additional HTTP server themselves.